### PR TITLE
Bridge lemma for FragmentId ordering; Verus proof for CandidatePair::new (7.2.6)

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Introduces a Verus bridge lemma for FragmentId ordering and wires it into a dedicated Verus proof for CandidatePair::new canonicalization. Extracts FragmentId into its own module to model ordering independently from String internals. Adds unit tests and behaviour-driven tests for the constructor seam, plus a Verus sidecar proof and runner wiring. Documentation and roadmap updates accompany the changes.

## Changes
- New/updated files:
  - crates/whitaker_clones_core/src/index/fragment_id.rs
  - crates/whitaker_clones_core/src/index/mod.rs
  - crates/whitaker_clones_core/src/index/types.rs
  - crates/whitaker_clones_core/src/index/tests.rs
  - crates/whitaker_clones_core/tests/candidate_pair_behaviour.rs
  - crates/whitaker_clones_core/tests/features/candidate_pair.feature
  - verus/clone_detector_candidate_pair.rs
  - scripts/run-verus.sh (wired in the new proof file)
  - docs/execplans/7-2-6-verus-proofs-for-candidate-pair-new-canonicalization.md
  - docs/whitaker-clone-detector-design.md (updated with 7.2.6 decisions)
  - docs/developers-guide.md (updated to reflect Verus workflow for 7.2.6)
  - docs/roadmap.md (7.2.6 marked done)

## Rationale
- Documents and enforces two core invariants for CandidatePair::new:
  1) Distinct fragment identifiers are returned in canonical lexical order regardless of input order.
  2) Identical fragment identifiers yield None, suppressing self-pairs at the constructor seam.
- Establishes a formal bridge to a lexically ordered FragmentId domain via a Verus proof, keeping proof-shape aligned with ADR003 and existing clone-detector proof patterns.
- Introduces a dedicated FragmentId bridge lemma so the Verus proof does not rely on internal String semantics, while runtime tests pin the concrete lexical behavior.

## Implementation plan (status)
- Stage B: Direct unit tests for CandidatePair::new (happy/unhappy/edge cases) – implemented in tests.rs.
- Stage C: Behaviour-driven tests (rstest-bdd) for constructor semantics – implemented in candidate_pair_behaviour.rs and candidate_pair.feature.
- Stage D: Dedicated Verus sidecar proof file wired into the verifier workflow – implemented in verus/clone_detector_candidate_pair.rs.
- Stage E: Wire the new proof into the repository’s verus runner so clone-detector and all groups run it – wired in scripts/run-verus.sh.
- Stage F: Update design/developer docs to reflect 7.2.6 decisions – reflected in updated docs.
- Stage G: Mark 7.2.6 done in docs/roadmap.md after gates pass – updated.
- Stage H: Validate via fmt, lint, tests, and Verus runs; logs captured with pipefail – validated via repository tests and verus runs.

## Validation plan
- Run the repository’s full validation sequence as guided in the ExecPlan:
  - make fmt, make markdownlint, make nixie, make check-fmt, make lint, make test
  - make verus-clone-detector, and make verus
- Expect direct CandidatePair::new tests and BDD scenarios to pass, with Verus proofs succeeding alongside existing clone-detector proofs.
- Update docs/roadmap.md to reflect 7.2.6 completion once gates succeed.

## Documentation impact
- The new ExecPlan documents the final proof scope, ordering contract, and trust boundary for the clone-detector proofs.
- No public API changes or user-facing behavior are introduced by this documentation alone.

## Review guidance
- Focus on scope adherence to 7.2.6 and the clarity of the stated invariants.
- Ensure the plan aligns with ADR 003 and existing clone-detector proof patterns.
- Confirm that the staged approach minimizes drift and keeps changes modular.

> Generated by DevBoxer; task reference: d830efa2-6dc2-4e41-806e-b81a271753bb

📎 **Task**: https://www.devboxer.com/task/d830efa2-6dc2-4e41-806e-b81a271753bb

📎 **Task**: https://www.devboxer.com/task/316de405-a8f6-4f96-99aa-72c5e69a2865

## Summary by Sourcery

Document and formally verify the `CandidatePair::new` constructor contract, including canonical fragment ordering and self-pair suppression, while extracting `FragmentId` into its own module and wiring the new Verus proof into the existing proof workflow.

New Features:
- Add a dedicated Verus sidecar proof for the `CandidatePair::new` constructor and integrate it into the clone-detector and full Verus proof groups.
- Introduce a behaviour-driven test harness and feature file to describe `CandidatePair::new` canonicalization and self-pair suppression scenarios.

Enhancements:
- Extract `FragmentId` into its own module and re-export it from the index module to decouple its definition from other index types.
- Add unit tests that directly cover `CandidatePair::new` ordering behaviour and self-pair suppression, including a lexical ordering edge case.
- Clarify design and developer documentation around clone-detector constructor proofs, the `FragmentId` ordering contract, and the Verus trust boundary, and mark roadmap item 7.2.6 as complete.

Documentation:
- Add an execution plan documenting the scope, rationale, and outcomes for the `CandidatePair::new` Verus proof and related tests.
- Update clone-detector design and developer guides to reflect the new constructor proof, ordering contract, and proof workflow, and mark roadmap item 7.2.6 as done.

Tests:
- Add focused unit tests and BDD scenarios for `CandidatePair::new` to validate canonical ordering and self-pair suppression at the constructor seam.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai are you satisfied that the introduced proof is substantive, rigorous, and well-founded?

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue.

### ❌ Failed checks (1 error, 1 warning, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Module-Level Documentation | ❌ Error | The verus/clone\_detector\_candidate\_pair.rs file contains four undocumented public specification functions and one undocumented public struct, along with two undocumented assume\_specification declarations. | Add documentation comments to all undocumented public declarations in verus/clone\_detector\_candidate\_pair.rs, including ExFragmentId, fragment\_id\_rank, fragment\_id\_partial\_cmp\_model, strict\_total\_order\_axioms, and both assume\_specification items. |
| Developer Documentation | ⚠️ Warning | Internal architectural change extracting FragmentId into dedicated module is not explicitly documented in developers-guide.md as required. | Add section to developers-guide.md explaining the internal structure of crates/whitaker\_clones\_core/src/index/ and the rationale for FragmentId module reorganisation. |
| Testing (Unit) | ❓ Inconclusive | Test files claimed in PR summary (tests.rs, candidate\_pair\_behaviour.rs, candidate\_pair.feature) cannot be located in the repository. | Verify test files are committed and present. Confirm unit tests cover edge cases (self-pairs, lexical ordering, reversed inputs). Validate BDD scenarios exercise functional boundaries, not internal helpers. Ensure testing dependencies (rstest, cucumber) are declared in Cargo.toml. |

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- In the BDD harness, `CandidatePairWorld` stores `left` and `right` as `String` and re-wraps them into `FragmentId` in the `when` step; consider storing `FragmentId` directly to avoid unnecessary conversions and keep the test world aligned with the production types.
- In the Verus sidecar, `fragment_id_rank` is used as the basis for equality and ordering but its intended properties (e.g., that equal ranks correspond exactly to `eq_spec`, and that different ranks imply inequality) are only encoded indirectly in the lemmas; you could add a brief spec comment or helper lemma making those assumptions explicit to clarify the bridge contract for future readers.

## Individual Comments

### Comment 1
<location path="docs/developers-guide.md" line_range="150-155" />
<code_context>
+- Verus proves the `LshConfig::new` constructor model rejects zero bands,
+  rejects zero rows, accepts only exact products of `MINHASH_SIZE`, and rejects
+  overflowing products via the same `checked_mul` semantics as the runtime code.
+- Verus proves the `CandidatePair::new` constructor model suppresses equal
+  inputs, preserves already ordered distinct inputs, and swaps reversed
+  distinct inputs into canonical order. The sidecar now states that proof over
+  a trusted `FragmentId` bridge lemma: `FragmentId::partial_cmp` is modelled as
+  a strict total order via a ghost `nat` ranking, rather than being left as an
+  implicit assumption.
 - Kani executes the real constructor with one concrete acceptance harness, one
</code_context>
<issue_to_address>
**suggestion (typo):** Clarify the phrase "states that proof over" for grammatical correctness.

The sentence reads as ungrammatical. Consider rephrasing to something like "The sidecar now states that proof as being over a trusted `FragmentId` bridge lemma" or "The sidecar now formulates that proof over a trusted `FragmentId` bridge lemma" to improve clarity.

```suggestion
- Verus proves the `CandidatePair::new` constructor model suppresses equal
  inputs, preserves already ordered distinct inputs, and swaps reversed
  distinct inputs into canonical order. The sidecar now formulates that proof
  over a trusted `FragmentId` bridge lemma: `FragmentId::partial_cmp` is
  modelled as a strict total order via a ghost `nat` ranking, rather than being
  left as an implicit assumption.
```
</issue_to_address>

### Comment 2
<location path="verus/clone_detector_candidate_pair.rs" line_range="82" />
<code_context>
+    right: &FragmentId,
+) -> Option<Ordering>;
+
+pub open spec fn strict_total_order_axioms<Id: PartialOrd>() -> bool {
+    &&& forall|id: Id| #[trigger] id.eq_spec(&id)
+    &&& forall|left: Id, right: Id| #[trigger] left.partial_cmp_spec(&right)
</code_context>
<issue_to_address>
**issue (complexity):** Consider specializing the proof to `FragmentId`, inlining thin spec layers, and reusing a helper lemma so the same behavior and trust boundary are expressed with fewer generics, wrappers, and duplicated case splits.

You can keep all the behaviour and the trust boundary while substantially reducing complexity by specializing the proof to `FragmentId` and collapsing some layers. Three focused refactors:

---

### 1. Specialize the order axioms to `FragmentId`

The generic `strict_total_order_axioms<Id: PartialOrd>()` and its generic lemmas only ever get instantiated with `FragmentId`. You can avoid the generic parameter and the associated type noise by defining a `FragmentId`‑specific axiom block and using it directly.

```rust
pub open spec fn fragment_id_strict_total_order_axioms() -> bool {
    &&& forall|id: FragmentId| #[trigger] id.eq_spec(&id)
    &&& forall|left: FragmentId, right: FragmentId| #[trigger]
        left.partial_cmp_spec(&right) != None::<Ordering>
    &&& forall|left: FragmentId, right: FragmentId| #[trigger]
        left.partial_cmp_spec(&right) == Some(Ordering::Equal)
            <==> left.eq_spec(&right)
    &&& forall|left: FragmentId, right: FragmentId| #[trigger]
        left.partial_cmp_spec(&right) == Some(Ordering::Less)
            <==> right.partial_cmp_spec(&left) == Some(Ordering::Greater)
    &&& forall|left: FragmentId, right: FragmentId| #[trigger]
        left.partial_cmp_spec(&right) == Some(Ordering::Greater)
            <==> right.partial_cmp_spec(&left) == Some(Ordering::Less)
    &&& forall|left: FragmentId, middle: FragmentId, right: FragmentId|
        left.partial_cmp_spec(&middle) == Some(Ordering::Less) && #[trigger]
            middle.partial_cmp_spec(&right) == Some(Ordering::Less) ==> #[trigger]
            left.partial_cmp_spec(&right) == Some(Ordering::Less)
}
```

Then your constructor spec and lemmas no longer need a type parameter:

```rust
spec fn candidate_pair_new_result(
    left: FragmentId,
    right: FragmentId,
) -> CandidatePairOutcome<FragmentId> {
    if left.eq_spec(&right) {
        CandidatePairOutcome::NoPair
    } else if left.partial_cmp_spec(&right) == Some(Ordering::Less) {
        CandidatePairOutcome::Pair(left, right)
    } else {
        CandidatePairOutcome::Pair(right, left)
    }
}
```

This removes all `Id: PartialOrd` generics from the core proof while preserving the same logical content.

---

### 2. Collapse the layered spec API for the constructor contract

`constructor_accepts` and `is_canonical_pair` are only thin wrappers around `candidate_pair_new_result` and are only used in the proof tree. You can inline them into the main contract lemma to remove several layers and almost‑empty lemmas.

Instead of three separate specs:

```rust
spec fn constructor_accepts<Id: PartialOrd>(left: Id, right: Id) -> bool { ... }
spec fn is_canonical_pair<Id: PartialOrd>(left: Id, right: Id, pair_left: Id, pair_right: Id) -> bool { ... }

proof fn lemma_constructor_contract<Id: PartialOrd>(...) { ... }
```

You can keep just `candidate_pair_new_result` and express the contract directly:

```rust
proof fn lemma_fragment_id_constructor_contract(left: FragmentId, right: FragmentId)
    requires
        fragment_id_strict_total_order_axioms(),
    ensures
        match candidate_pair_new_result(left, right) {
            CandidatePairOutcome::NoPair =>
                left.eq_spec(&right),
            CandidatePairOutcome::Pair(pair_left, pair_right) =>
                // canonical pair: ordered and is a permutation of inputs
                pair_left.partial_cmp_spec(&pair_right) == Some(Ordering::Less)
                && ((pair_left.eq_spec(&left) && pair_right.eq_spec(&right))
                    || (pair_left.eq_spec(&right) && pair_right.eq_spec(&left))),
        },
{
    // Single case split proves full contract:
    match left.partial_cmp_spec(&right) {
        Some(Ordering::Less) => {
            // equal case ruled out by eq_spec <==> Equal axiom
            assert(!left.eq_spec(&right));
        }
        Some(Ordering::Equal) => {
            assert(left.eq_spec(&right));
        }
        Some(Ordering::Greater) => {
            assert(!left.eq_spec(&right));
        }
        None => {
            assert(false); // impossible under axioms
        }
    }
}
```

If you still want a boolean “accepts” view for callers, you can define it as a *pure alias* of the pattern match, but keep it out of the proof structure:

```rust
spec fn constructor_accepts(left: FragmentId, right: FragmentId) -> bool {
    matches!(candidate_pair_new_result(left, right), CandidatePairOutcome::Pair(_, _))
}
```

This removes `constructor_accepts`, `is_canonical_pair`, and the intermediate lemmas from the core proof path without changing the externally visible contract.

---

### 3. Factor the repeated rank‑based case splits into a helper lemma

`lemma_fragment_id_partial_cmp_obeys_strict_total_order` repeats the same `< / == / >` case analysis over `fragment_id_rank` in multiple `forall` blocks. You can encapsulate this once in a small helper that relates ranks to `partial_cmp_spec`, and then reuse it.

```rust
proof fn lemma_partial_cmp_matches_rank(left: FragmentId, right: FragmentId)
    ensures
        if fragment_id_rank(&left) < fragment_id_rank(&right) {
            left.partial_cmp_spec(&right) == Some(Ordering::Less)
                && right.partial_cmp_spec(&left) == Some(Ordering::Greater)
        } else if fragment_id_rank(&left) > fragment_id_rank(&right) {
            left.partial_cmp_spec(&right) == Some(Ordering::Greater)
                && right.partial_cmp_spec(&left) == Some(Ordering::Less)
        } else {
            left.partial_cmp_spec(&right) == Some(Ordering::Equal)
                && right.partial_cmp_spec(&left) == Some(Ordering::Equal)
        },
{
    if fragment_id_rank(&left) < fragment_id_rank(&right) {
        assert(left.partial_cmp_spec(&right) == Some(Ordering::Less));
        assert(right.partial_cmp_spec(&left) == Some(Ordering::Greater));
    } else if fragment_id_rank(&left) > fragment_id_rank(&right) {
        assert(left.partial_cmp_spec(&right) == Some(Ordering::Greater));
        assert(right.partial_cmp_spec(&left) == Some(Ordering::Less));
    } else {
        assert(left.partial_cmp_spec(&right) == Some(Ordering::Equal));
        assert(right.partial_cmp_spec(&left) == Some(Ordering::Equal));
    }
}
```

Then the main axiom lemma can reuse this instead of re‑doing the case split each time:

```rust
proof fn lemma_fragment_id_partial_cmp_obeys_strict_total_order()
    ensures
        fragment_id_strict_total_order_axioms(),
{
    assert forall|id: FragmentId| #[trigger] id.eq_spec(&id) by { /* trivial */ };

    assert forall|left: FragmentId, right: FragmentId|
        #[trigger] left.partial_cmp_spec(&right) != None::<Ordering> by {
        lemma_partial_cmp_matches_rank(left, right);
    };

    assert forall|left: FragmentId, right: FragmentId|
        #[trigger] left.partial_cmp_spec(&right) == Some(Ordering::Equal)
            <==> left.eq_spec(&right) by {
        lemma_partial_cmp_matches_rank(left, right);
    };

    assert forall|left: FragmentId, right: FragmentId|
        #[trigger] left.partial_cmp_spec(&right) == Some(Ordering::Less)
            <==> right.partial_cmp_spec(&left) == Some(Ordering::Greater) by {
        lemma_partial_cmp_matches_rank(left, right);
    };

    // ...similar reuse for the > / < equivalence and transitivity...
}
```

This keeps the ranking model and trust boundary intact but eliminates the duplicated branching logic.

---

These changes keep all the logical guarantees and the explicit trust boundary while reducing genericity, removing thin abstraction layers, and de‑duplicating rank‑based reasoning, making the proof much easier to read and maintain.
</issue_to_address>

### Comment 3
<location path="verus/clone_detector_candidate_pair.rs" line_range="39" />
<code_context>
+
+pub uninterp spec fn fragment_id_rank(id: &FragmentId) -> nat;
+
+pub open spec fn fragment_id_partial_cmp_model(
+    left: &FragmentId,
+    right: &FragmentId,
</code_context>
<issue_to_address>
**issue (review_instructions):** `fragment_id_partial_cmp_model` uses an `if / else if / else` chain (>2 branches) directly instead of delegating to a predicate spec function as required.

Within `fragment_id_partial_cmp_model`, the `if fragment_id_rank(left) < ... { } else if ... { } else { }` structure has three branches. Per the instructions, conditionals with more than two branches should be factored into a separate predicate function (e.g. a small spec that returns the `Ordering`), which this function can then call.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>

### Comment 4
<location path="verus/clone_detector_candidate_pair.rs" line_range="180" />
<code_context>
+    };
+}
+
+spec fn candidate_pair_new_result<Id: PartialOrd>(left: Id, right: Id) -> CandidatePairOutcome<Id> {
+    if left.eq_spec(&right) {
+        CandidatePairOutcome::NoPair
</code_context>
<issue_to_address>
**issue (review_instructions):** `candidate_pair_new_result` implements a three-way branch directly instead of delegating the branching logic to a predicate function.

`candidate_pair_new_result` currently uses a three-way `if left.eq_spec(&right) { } else if left.partial_cmp_spec(&right) == Some(Ordering::Less) { } else { }`. That is a >2 branch conditional; per the instruction, this logic should be factored into a dedicated predicate/spec helper (e.g. returning an enum describing the relation) that this function then maps to `CandidatePairOutcome`.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>

### Comment 5
<location path="verus/clone_detector_candidate_pair.rs" line_range="235" />
<code_context>
+    assert(right.partial_cmp_spec(&left) == Some(Ordering::Less));
+}
+
+proof fn lemma_distinct_inputs_yield_one_canonical_pair<Id: PartialOrd>(left: Id, right: Id)
+    requires
+        strict_total_order_axioms::<Id>(),
</code_context>
<issue_to_address>
**issue (review_instructions):** `lemma_distinct_inputs_yield_one_canonical_pair` performs a four-way `match` on `partial_cmp_spec`, which exceeds the allowed number of branches for inlined conditionals.

In `lemma_distinct_inputs_yield_one_canonical_pair`, the `match left.partial_cmp_spec(&right) { Some(Less) => ..., Some(Equal) => ..., Some(Greater) => ..., None => ... }` has four branches. The style rule requires conditionals with more than two branches to be moved into a predicate/helper; consider introducing a small spec/predicate over the comparison result and using that here, keeping the lemma focused on its proof obligations rather than the branching structure.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.rs`

**Instructions:**
Move conditionals with >2 branches into a predicate function.

</details>
</issue_to_address>
~~~

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue.

### ❌ Failed checks (1 error, 1 warning, 1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Module-Level Documentation | ❌ Error | The verus/clone\_detector\_candidate\_pair.rs file contains four undocumented public specification functions and one undocumented public struct, along with two undocumented assume\_specification declarations. | Add documentation comments to all undocumented public declarations in verus/clone\_detector\_candidate\_pair.rs, including ExFragmentId, fragment\_id\_rank, fragment\_id\_partial\_cmp\_model, strict\_total\_order\_axioms, and both assume\_specification items. |
| Developer Documentation | ⚠️ Warning | Internal architectural change extracting FragmentId into dedicated module is not explicitly documented in developers-guide.md as required. | Add section to developers-guide.md explaining the internal structure of crates/whitaker\_clones\_core/src/index/ and the rationale for FragmentId module reorganisation. |
| Testing (Unit) | ❓ Inconclusive | Test files claimed in PR summary (tests.rs, candidate\_pair\_behaviour.rs, candidate\_pair.feature) cannot be located in the repository. | Verify test files are committed and present. Confirm unit tests cover edge cases (self-pairs, lexical ordering, reversed inputs). Validate BDD scenarios exercise functional boundaries, not internal helpers. Ensure testing dependencies (rstest, cucumber) are declared in Cargo.toml. |

**Comment by leynos** [comment]
@coderabbitai review

File: verus/clone_detector_candidate_pair.rs [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+
+use core::cmp::Ordering;
+
+use vstd::prelude::*;
+use vstd::std_specs::cmp::{PartialEqSpec, PartialOrdSpec};
+
+#[path = "../crates/whitaker_clones_core/src/index/fragment_id.rs"]
+mod fragment_id_runtime;
+
+use fragment_id_runtime::FragmentId;
```
[the remark this answers is not in the document]
Comment (leynos) [reply verus/clone_detector_candidate_pair.rs:26]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Unit And Behavioural) | ❓ Inconclusive | Unable to verify FragmentId testing coverage without access to actual codebase or verification output from the provided shell commands. | Execute the provided shell commands against the codebase to determine which public methods and traits are exported, identify any inline doctests, and check documentation for testing coverage assertions. |

**Comment by leynos** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

🚥 Pre-merge checks | ✅ 17 | ❌ 1

### ❌ Failed checks (1 inconclusive)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Unit And Behavioural) | ❓ Inconclusive | Unable to verify FragmentId testing coverage without access to actual codebase or verification output from the provided shell commands. | Execute the provided shell commands against the codebase to determine which public methods and traits are exported, identify any inline doctests, and check documentation for testing coverage assertions. |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

### ⚠️ Narrow gap — `Display` and `AsRef<str>`

Two trait impls have no test coverage of any kind:

| Item | Status |
| :-- | :-- |
| `impl fmt::Display for FragmentId` | No unit test, no doctest |
| `impl AsRef<str> for FragmentId` | No unit test, no doctest |

Both are single-line delegations to `as_str`, so the risk is low. However, given that you have asked that warnings not be treated as optional, I recommend opening a follow-up issue rather than leaving this untracked. Would you like me to open a GitHub issue for this?
